compiler/optimizer/SinkStores.cpp

```diff
@@ -2460,6 +2460,16 @@ bool TR_SinkStores::treeIsSinkableStore(TR::Node *node, bool sinkIndirectLoads,
             traceMsg(comp(), "         can't move store of pinning array reference or with UseOnlyAliases\n");
          return false;
          }
+#ifdef J9_PROJECT_SPECIFIC
+      if (node->getOpCode().isStore() && node->getFirstChild()->getOpCode().isBinaryCodedDecimalOp() &&
+              node->getFirstChild()->getOpCode().isConversion() &&
+             (node->getFirstChild()->getOpCode().getDataType() == TR::Int32 || node->getFirstChild()->getOpCode().getDataType() == TR::Int64))
```

Comment:
I'm not a fan of introducing additional `J9_PROJECT_SPECIFIC` ifdefs into OMR. @0xdaryl FYI. I see that `node->dontEliminateStores()` is called a few lines above. This is implemented here:

https://github.com/eclipse/omr/blob/8bbe28cea066becdd271db94888c331ae6a4601d/compiler/il/OMRNode.cpp#L2306-L2316

It seems to query `OMRSymbol` which is an extensible class:

https://github.com/eclipse/omr/blob/8bbe28cea066becdd271db94888c331ae6a4601d/compiler/il/symbol/OMRSymbol.hpp#L70-L77

So that begs the question: Could we do this in the upstream OpenJ9 project without introducing additional `J9_PROJECT_SPECIFIC` into the OMR codebase?

Reply:
In addition it seems this change may be too strict. Rather than checking for `TR::Int32` and `TR::Int64` explicitly we should check simply that the node is of an integral type since this is the case you really care about. I'd also like to see a comment describing why this restriction is being put in place directly in the code. i.e. describe why sinking integral stores for packed operations which could potentially be under a BCDCHK operation is not safe.

Reply:
+1

Reply:
@fjeremic Yes, this would work better. I think the best place we can do this in the `Node` class. I think we should be able to override `OMR::Node::dontEliminateStores` and have a `J9::Node::dontEliminateStores`. Then something like the following will probably work:
https://github.com/dchopra001/openj9/blob/f99987b1293c2bf4fc804b0bea03e5511b4a62f6/runtime/compiler/il/J9Node.cpp#L131-L138